File: 101_Symmetric_Tree/101_1.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True;
        return self.isSymmetricLeftRight(root.left, root.right)

    def isSymmetricLeftRight(self, left: TreeNode, right: TreeNode):
        if not left and not right:
            return True
        if not left or not right:
           return False
        return left.val == right.val and \
            self.isSymmetricLeftRight(left.left, right.right) and \
            self.isSymmetricLeftRight(left.right, right.left)


if __name__ == "__main__":
    a = TreeNode(1)
    b = TreeNode(2)
    c = TreeNode(2)
    d = TreeNode(3)
    e = TreeNode(4)
    f = TreeNode(4)
    g = TreeNode(3)
    a.left = b
    a.right = c
    b.left = d
    b.right = e
    c.left = f
    c.right = g
    s = Solution()
    s.isSymmetric(a)

File: 101_Symmetric_Tree/101_1.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {   // recursion
public:
    static bool isSymmetric(TreeNode* root) {
        if (!root)  return true;
        return isSymmetricLeftRight(root->left, root->right);
    }

    static bool isSymmetricLeftRight(TreeNode *left, TreeNode *right) {
        if (!left && !right) return true;
        if (!left || !right) return false;
        return (left->val == right->val) && 
                isSymmetricLeftRight(left->left, right->right) && 
                isSymmetricLeftRight(left->right, right->left);
    }
};

int main() {
    TreeNode *a = new TreeNode(1);
    TreeNode *b = new TreeNode(2);
    TreeNode *c = new TreeNode(2);
    TreeNode *d = new TreeNode(3);
    TreeNode *e = new TreeNode(4);
    TreeNode *f = new TreeNode(4);
    TreeNode *g = new TreeNode(3);
    a->left = b;
    a->right = c;
    b->left = d;
    b->right = e;
    c->left = f;
    c->right = g;
    cout << Solution::isSymmetric(a) << endl;
}